README.md
# Windows Imaging File Format (WIM)
[![npm](https://img.shields.io/npm/v/wim.svg?style=flat-square)](https://npmjs.com/package/wim)
[![npm license](https://img.shields.io/npm/l/wim.svg?style=flat-square)](https://npmjs.com/package/wim)
[![npm downloads](https://img.shields.io/npm/dm/wim.svg?style=flat-square)](https://npmjs.com/package/wim)
[![build status](https://img.shields.io/travis/jhermsmeier/node-wim.svg?style=flat-square)](https://travis-ci.org/jhermsmeier/node-wim)

## Install via [npm](https://npmjs.com)

```sh
$ npm install --save wim
```

## References

- [Windows Imaging File Format on TechNet](https://technet.microsoft.com/en-us/library/cc749478(v=ws.10).aspx)
- [Windows Imaging File Format (WIM) white paper](http://go.microsoft.com/fwlink/?LinkId=92227)

## Usage

```js
var WIM = require( 'wim' )
```

### Read a WIM header

```js
var image = new WIM.Image()

image.open( 'sources.wim', function( error ) {
  if( error ) return handleError( error )
  image.readHeader( function( error, header ) {
    console.log( header )
  })
})
```

```js
Header {
  imageTag: 'MSWIM\u0000\u0000\u0000',
  size: 208,
  version: 68864,
  flags: 262274,
  compressedSize: 32768,
  guid: <Buffer 29 f7 36 06 03 77 e4 41 96 f8 1b b8 69 fd 49 7d>,
  partNumber: 1,
  partCount: 1,
  imageCount: 2,
  offsetTable: FileHeader {
     size: 461850,
     flags: 2,
     offset: 307668199,
     originalSize: 461850
  },
  xmlData: FileHeader {
    size: 3824,
    flags: 2,
    offset: 308130049,
    originalSize: 3824
  },
  bootMetadata: FileHeader {
     size: 920042,
     flags: 6,
     offset: 306748157,
     originalSize: 4555704
  },
  bootIndex: 2,
  integrity: FileHeader {
    size: 0,
    flags: 0,
    offset: 0,
    originalSize: 0
  },
  reserved: <Buffer 00 00 00 00 00 00 00 00 00 00 00 00 00  ... >
}
```

### CLI Usage
```
npx wim wimfile -h|-x|-m
  dump wimfile header, xml data or metadata
```

### CLI - read version from windows installation ISO 

ISO mount on filesystem is Linux-specific here.

```
mount win10.iso /mnt/win10/ -o loop
npx wim /mnt/win10/sources/install.wim -x | npx --package @toycode/xml2json-cli xml2json | jq -r .WIM.IMAGE[0].WINDOWS[0].SERVICINGDATA[0].PKEYCONFIGVERSION[0]
```

You get `10.0.19041.1202;2016-01-01T00:00:00Z` for instance.


CHANGELOG.md
<a name="0.1.0"></a>
# 0.2.0 (2022-06-25)

### Features

* Adding cli to dump header, XML data or boot metadata ([0bfbbf7](https://github.com/jhermsmeier/node-wim/commit/0bfbbf7)) [[Frederic Germain](https://github.com/fredericgermain)]

### Fixes

* Fix Buffer API on newer nodejs (14, 16) ([08ce7e7](https://github.com/jhermsmeier/node-wim/commit/08ce7e7)) [[Frederic Germain](https://github.com/fredericgermain)]

<a name="0.1.0"></a>
# 0.1.0 (2017-02-27)

### Features

* **wim:** Implement parsing various structures ([e7b67f3](https://github.com/jhermsmeier/node-wim/commit/e7b67f3))


test/wim.js
var WIM = require( '..' )
var assert = require( 'assert' )
var path = require( 'path' )
var util = require( 'util' )

function inspect( value ) {
  return util.inspect( value, {
    depth: null,
    colors: process.stdout.isTTY,
  })
}

function log( value ) {
  console.log( '' )
  console.log( inspect( value ) )
  console.log( '' )
}

before( function() {
  log( WIM )
})

describe( 'WIM', function() {

  context( 'Image', function() {

    // var WIM_IMAGE_PATH = path.resolve( 'E:/sources/boot.wim' )
    var WIM_IMAGE_PATH = path.join( __dirname, 'data', 'boot-header.wim' )
    var image = null

    specify( 'new', function() {
      image = new WIM.Image()
    })

    specify( '#open()', function( done ) {
      image.open( WIM_IMAGE_PATH, done )
    })

    specify( '#readHeader()', function( done ) {
      image.readHeader( function( error, header ) {
        log( error || header )
        done( error )
      })
    })

    specify.skip( '#readMetadata()', function( done ) {
      image.readMetadata( function( error, metadata ) {
        log( error || metadata )
        done( error )
      })
    })

    specify.skip( '#readXmlData()', function( done ) {
      image.readXmlData( function( error, metadata ) {
        log( error || metadata )
        done( error )
      })
    })

    specify( '#close()', function( done ) {
      image.close( done )
    })

  })

})


cli.js
#!/usr/bin/env node

var WIM = require( './lib/wim' )
var image = new WIM.Image()
 
if (process.argv.length < 4) {
    console.log('npx wim wimfile -h|-x|-m')
    console.log('  dump wimfile header, xml data or metadata')
    process.exit(-1)
}

function handleError(error) {
    console.error(error);
    process.exit(-1)
}

image.open( process.argv[2], function( error ) {
  if( error ) return handleError( error )
  image.readHeader( function( error, header ) {
    if (process.argv[3] === '-h') {
      console.log( header )
    }
    if (process.argv[3] === '-x') {
        image.readXmlData( function( error, header ) {
            console.log( header )
        })
    }
    if (process.argv[3] === '-m') {
        image.readMetadata( function( error, header ) {
            console.log( header )
        })
    }
  })
})


lib/wim.js
var WIM = module.exports

WIM.HEADER_SIZE = 208

// Header flags
WIM.RESERVED = 0x00000001
WIM.COMPRESSION = 0x00000002
WIM.READONLY = 0x00000004
WIM.SPANNED = 0x00000008
WIM.RESOURCE_ONLY = 0x00000010
WIM.METADATA_ONLY = 0x00000020
WIM.WRITE_IN_PROGRESS = 0x00000040
WIM.REPARSE_FIX = 0x00000080
WIM.COMPRESS_RESERVED = 0x00010000
WIM.COMPRESS_XPRESS = 0x00020000
WIM.COMPRESS_LZX = 0x00040000

// File Resource Header flags
WIM.FILE_FREE = 0x01
WIM.FILE_METADATA = 0x02
WIM.FILE_COMPRESSED = 0x04
WIM.FILE_SPANNED = 0x08

WIM.extensions = [ '.wim', '.swm' ]

WIM.Header = require( './header' )
WIM.SecurityData = require( './secdata' )
WIM.DirectoryEntry = require( './direntry' )
WIM.FileResource = require( './file-resource' )
WIM.Image = require( './image' )

/**
 * Open a WIM format image
 * @param {String} filename
 * @param {Object} [options]
 * @param {Object} [options.fd] - file descriptor
 * @param {Object} [options.fs] - user `fs` api
 * @param {Function} callback - 'open' event callback
 * @returns {WIM.Image}
 */
WIM.open = function( filename, options, callback ) {
  return new WIM.Image().open( filename, options, callback )
}


lib/image.js
var WIM = require( './wim' )

/**
 * Image
 * @constructor
 * @memberOf WIM
 * @param {String} filename
 * @param {Object} [options]
 * @returns {Image}
 */
function Image( filename, options ) {

  if( !(this instanceof Image) )
    return new Image( filename, options )

  options = options || {}

  this.fs = options.fs || require( 'fs' )
  this.filename = filename
  this.flags = options.flags || 'r'
  this.fd = options.fd || null

  this.header = new WIM.Header()
  this.securityData = new WIM.SecurityData()
  this.xmlData = null

  Object.defineProperty( this, 'fs', {
    enumberable: false
  })

}

/**
 * Image prototype
 * @type {Object}
 * @ignore
 */
Image.prototype = {

  constructor: Image,

  readXmlData( callback ) {

    var buffer = Buffer.alloc( this.header.xmlData.size, 0 )
    var length = buffer.length
    var position = this.header.xmlData.offset

    if( !this.header.xmlData.getFlag( WIM.FILE_METADATA ) ) {
      return callback.call( this, new Error( 'Not a metadata file' ) )
    }

    this.fs.read( this.fd, buffer, 0, length, position, ( error, bytesRead, buffer ) => {
      if( this.header.xmlData.getFlag( WIM.FILE_COMPRESSED ) ) {
        return callback.call( this, new Error( 'Compressed file resources not implemented' ) )
      }
      this.xmlData = buffer.toString( 'utf16le' )
      callback.call( this, error, this.xmlData )
    })

  },

  readMetadata( callback ) {

    var buffer = Buffer.alloc( this.header.bootMetadata.size, 0 )
    var length = buffer.length
    var position = this.header.bootMetadata.offset

    if( !this.header.bootMetadata.getFlag( WIM.FILE_METADATA ) ) {
      return callback.call( this, new Error( 'Not a metadata file' ) )
    }

    this.fs.read( this.fd, buffer, 0, length, position, ( error, bytesRead, buffer ) => {
      if( this.header.bootMetadata.getFlag( WIM.FILE_COMPRESSED ) ) {
        return callback.call( this, new Error( 'Compressed file resources not implemented' ) )
      }
      callback.call( this, error, this.securityData )
    })

  },

  readHeader( callback ) {

    var buffer = Buffer.alloc( WIM.HEADER_SIZE, 0 )
    var length = buffer.length
    var position = 0

    this.fs.read( this.fd, buffer, 0, length, position, ( error, bytesRead, buffer ) => {
      this.header.parse( buffer )
      callback.call( this, error, this.header )
    })

  },

  open( filename, options, callback ) {

    if( typeof options === 'function' ) {
      callback = options
      options = {}
    }

    // TODO: Close & reopen, or throw?
    if( this.fd != null ) {
      return callback.call( this, new Error( 'Already opened' ) )
    }

    if( options.fd ) {
      this.fd = options.fd
      return callback.call( this )
    }

    this.fs.open( filename, this.flags, ( error, fd ) => {
      this.fd = fd
      callback.call( this, error )
    })

    return this

  },

  close( callback ) {
    this.fs.close( this.fd, ( error ) => {
      callback.call( this, error )
    })
  },

}

module.exports = Image


lib/secdata.js
var WIM = require( './wim' )

/**
 * SecurityData
 * @constructor
 * @memberOf WIM
 * @returns {SecurityData}
 */
function SecurityData() {

  if( !(this instanceof SecurityData) )
    return new SecurityData()

  /** @type {Number} Total size of the descriptor table (in entries?) */
  this.tableLength = 0x00000000
  /** @type {Number} Number of directory entries */
  this.entryCount = 0x00000000
  /** @type {Array} Sizes of the actual descriptors */
  this.entrySizes = []
  /** @type {Array<WIM.DirectoryEntry>} Security descriptors */
  this.entries = []

}

/**
 * Parse a WIM SecurityData
 * @param {Buffer} buffer
 * @param {Number} [offset=0]
 * @returns {SecurityData}
 */
SecurityData.parse = function( buffer, offset ) {
  return new SecurityData().parse( buffer, offset )
}

/**
 * SecurityData prototype
 * @type {Object}
 * @ignore
 */
SecurityData.prototype = {

  constructor: SecurityData,

  parseEntries( buffer, offset ) {

    offset = offset || 0

    var entry = null

    for( var i = 0; i < this.entryCount; i++ ) {
      offset += this.entrySizes[i]
      entry = WIM.DirectoryEntry.parse( buffer, offset )
      this.entries.push( entry )
    }

    return this

  },

  parseHeader( buffer, offset ) {

    offset = offset || 0

    this.tableLength = buffer.readUInt32LE( offset + 0 )
    this.entryCount = buffer.readUInt32LE( offset + 4 )
    this.entrySizes = []

    for( var i = 0; i < this.tableLength; i++ ) {
      this.entrySizes.push( buffer.readUIntLE( offset + 8 + i * 8, 8 ) )
    }

    return this

  },

  /**
   * Parse WIM SecurityData
   * @param {Buffer} buffer
   * @param {Number} [offset=0]
   * @returns {SecurityData}
   */
  parse( buffer, offset ) {
    offset = offset || 0
    this.parseHeader( buffer, offset )
    return this
  },

}

module.exports = SecurityData


lib/direntry.js
/**
 * DirectoryEntry
 * @constructor
 * @memberOf WIM
 * @returns {DirectoryEntry}
 */
function DirectoryEntry() {

  if( !(this instanceof DirectoryEntry) )
    return new DirectoryEntry()

}

/**
 * Parse a WIM DirectoryEntry
 * @param {Buffer} buffer
 * @param {Number} [offset=0]
 * @returns {DirectoryEntry}
 */
DirectoryEntry.parse = function( buffer, offset ) {
  return new DirectoryEntry().parse( buffer, offset )
}

/**
 * DirectoryEntry prototype
 * @type {Object}
 * @ignore
 */
DirectoryEntry.prototype = {

  constructor: DirectoryEntry,

  /**
   * Parse a WIM DirectoryEntry
   * @param {Buffer} buffer
   * @param {Number} [offset=0]
   * @returns {DirectoryEntry}
   */
  parse( buffer, offset ) {

    offset = offset || 0

    return this

  },

}

module.exports = DirectoryEntry


lib/file-header.js
/**
 * FileHeader
 * @constructor
 * @memberOf WIM
 * @returns {FileHeader}
 */
function FileHeader() {

  if( !(this instanceof FileHeader) )
    return new FileHeader()

  // _RESHDR_BASE_DISK
  this.size = 0x00000000000000 // 7 bytes (!)
  this.flags = 0x00
  this.offset = 0x0000000000000000

  // _RESHDR_DISK_SHORT
  this.originalSize = 0x0000000000000000

}

FileHeader.size = 24

/**
 * Parse a WIM FileHeader
 * @param {Buffer} buffer
 * @param {Number} [offset=0]
 * @returns {FileHeader}
 */
FileHeader.parse = function( buffer, offset ) {
  return new FileHeader().parse( buffer, offset )
}

/**
 * FileHeader prototype
 * @type {Object}
 * @ignore
 */
FileHeader.prototype = {

  constructor: FileHeader,

  getFlag( mask ) {
    return ( this.flags & mask ) === mask
  },

  /**
   * Parse a WIM FileHeader
   * @param {Buffer} buffer
   * @param {Number} [offset=0]
   * @returns {FileHeader}
   */
  parse( buffer, offset ) {

    offset = offset || 0

    // _RESHDR_BASE_DISK
    this.size = buffer.readUIntLE( offset + 0, 6 ) + (1<<6)*buffer.readUIntLE( offset + 6, 1 )
    this.flags = buffer.readUInt8( offset + 7 )
    this.offset = Number(buffer.readBigInt64LE( offset + 8 ))

    // _RESHDR_DISK_SHORT
    this.originalSize = Number(buffer.readBigInt64LE( offset + 16 ))

    return this

  },

}

module.exports = FileHeader


lib/file-resource.js
/**
 * FileResource
 * @constructor
 * @memberOf WIM
 * @returns {FileResource}
 */
function FileResource() {

  if( !(this instanceof FileResource) )
    return new FileResource()

}

FileResource.Header = require( './file-header' )

/**
 * Parse a WIM FileResource
 * @param {Buffer} buffer
 * @param {Number} [offset=0]
 * @returns {FileResource}
 */
FileResource.parse = function( buffer, offset ) {
  return new FileResource().parse( buffer, offset )
}

/**
 * FileResource prototype
 * @type {Object}
 * @ignore
 */
FileResource.prototype = {

  constructor: FileResource,

  /**
   * Parse a WIM FileResource
   * @param {Buffer} buffer
   * @param {Number} [offset=0]
   * @returns {FileResource}
   */
  parse( buffer, offset ) {

    offset = offset || 0

    return this

  },

}

module.exports = FileResource


lib/header.js
var WIM = require( './wim' )

/**
 * Header
 * @constructor
 * @memberOf WIM
 * @returns {Header}
 */
function Header() {

  if( !(this instanceof Header) )
    return new Header()

  this.imageTag = 'MSWIM\0\0'
  this.size = 0x00000000
  this.version = 0x00000000
  this.flags = 0x00000000
  this.compressedSize = 0x00000000
  this.guid = Buffer.alloc(16)
  this.partNumber = 0x0001
  this.partCount = 0x0000
  this.imageCount = 0x00000000
  this.offsetTable = new WIM.FileResource.Header()
  this.xmlData = new WIM.FileResource.Header()
  this.bootMetadata = new WIM.FileResource.Header()
  this.bootIndex = 0x00000000
  this.integrity = new WIM.FileResource.Header()
  this.reserved = Buffer.alloc(60)

  this.guid.fill(0)
  this.reserved.fill(0)

}

Header.size = WIM.HEADER_SIZE

/**
 * Parse a WIM header
 * @param {Buffer} buffer
 * @param {Number} [offset=0]
 * @returns {Header}
 */
Header.parse = function( buffer, offset ) {
  return new Header().parse( buffer, offset )
}

/**
 * Header prototype
 * @type {Object}
 * @ignore
 */
Header.prototype = {

  constructor: Header,

  getFlag( mask ) {
    return ( this.flags & mask ) === mask
  },

  /**
   * Parse a WIM header
   * @param {Buffer} buffer
   * @param {Number} [offset=0]
   * @returns {Header}
   */
  parse( buffer, offset ) {

    offset = offset || 0

    this.imageTag = buffer.toString( 'ascii', offset + 0, offset + 8 )
    this.size = buffer.readUInt32LE( offset + 8 )
    this.version = buffer.readUInt32LE( offset + 12 )
    this.flags = buffer.readUInt32LE( offset + 16 )
    this.compressedSize = buffer.readUInt32LE( offset + 20 )

    buffer.copy( this.guid, 0, offset + 24, offset + 40 )

    this.partNumber = buffer.readUInt16LE( offset + 40 )
    this.partCount = buffer.readUInt16LE( offset + 42 )
    this.imageCount = buffer.readUInt32LE( offset + 44 )
    this.offsetTable.parse( buffer, offset + 48 )
    this.xmlData.parse( buffer, offset + 72 )
    this.bootMetadata.parse( buffer, offset + 96 )
    this.bootIndex = buffer.readUInt32LE( offset + 120 )
    this.integrity.parse( buffer, offset + 124 )

    buffer.copy( this.reserved, 0, offset + 148, offset + 208 )

    return this

  },

}

module.exports = Header
